sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -0,0 +1,274 @@
+  val foundPlans = new Array[mutable.Map[Set[Int], JoinPlan]](items.length + 1)
+  for (i <- 1 to items.length) foundPlans(i) = mutable.Map.empty
+
+  def search(): Option[LogicalPlan] = {
+    // Start from the first level: each plan is a single item with zero cost.
+    val itemIndex = items.zipWithIndex
+    foundPlans(1) ++=
+      itemIndex.map { case (item, id) => Set(id) -> JoinPlan(Set(id), item, Cost(0, 0)) }
+
+    for (lev <- 2 to items.length) {
+      searchForLevel(lev)
+    }
+
+    val plansLastLevel = foundPlans(items.length)
+    if (plansLastLevel.isEmpty) {
+      // Failed to find a plan, fall back to the original plan
+      None
+    } else {
+      // There must be only one plan at the last level, which contains all items.
+      assert(plansLastLevel.size == 1 && plansLastLevel.head._1.size == items.length)
+      Some(plansLastLevel.head._2.plan)
+    }
+  }
+
+  /** Find all possible plans in one level, based on previous levels. */
+  private def searchForLevel(level: Int): Unit = {
+    val foundPlansCurLevel = foundPlans(level)
+    var k = 1
+    var continue = true
+    while (continue) {
+      val otherLevel = level - k
+      if (k > otherLevel) {
+        // We can break from here, because when building a join from A and B, both A J B and B J A
+        // are handled.
+        continue = false
+      } else {
+        val joinPlansLevelK = foundPlans(k).values.toSeq
+        for (i <- joinPlansLevelK.indices) {
+          val curJoinPlan = joinPlansLevelK(i)
+
+          val joinPlansOtherLevel = if (k == otherLevel) {
+            // Both sides of a join are at the same level, no need to repeat for previous ones.
+            joinPlansLevelK.drop(i)
+          } else {
+            foundPlans(otherLevel).values.toSeq
+          }
+
+          joinPlansOtherLevel.foreach { otherJoinPlan =>
+            // Should not join two overlapping item sets.
+            if (curJoinPlan.itemIds.intersect(otherJoinPlan.itemIds).isEmpty) {
+              val joinPlan = buildJoin(curJoinPlan, otherJoinPlan)
+              if (joinPlan.nonEmpty) {
+                // Check if it's the first plan for the item set, or it's a better plan than
+                // the existing one due to lower cost.
+                val existingPlan = foundPlansCurLevel.get(joinPlan.get.itemIds)
+                if (existingPlan.isEmpty || joinPlan.get.cost < existingPlan.get.cost) {
+                  foundPlansCurLevel.update(joinPlan.get.itemIds, joinPlan.get)
+                }
+              }
+            }
+          }
+        }
+
+        k += 1
+      }
+    }
+  }
+
+  /** Build a new join node. */
+  private def buildJoin(curJoinPlan: JoinPlan, otherJoinPlan: JoinPlan): Option[JoinPlan] = {
+    // Check if these two nodes are inner joinable. We consider cartesian product very
+    // costly, thus exclude such plans. This also helps us to reduce the search space.
```

Comment:
After the `ReorderJoin` optimization rule, cross join are all at the end. Maybe I should use `Inner` join type instead of `InnerLike` while collecting joinable items, such that we don't need to worry about cross join during reordering?